cmd/relay/service.go

```diff
@@ -197,12 +198,25 @@ func (svc *Service) Shutdown() []error {
 }
 
 func (svc *Service) checkAdminAuth(next echo.HandlerFunc) echo.HandlerFunc {
-	headerVal := "Basic " + base64.StdEncoding.EncodeToString([]byte("admin:"+svc.config.AdminPassword))
+
+	// pre-compute valid HTTP auth headers based on the set of
+	validAuthHeaders := []string{}
+	for _, pw := range svc.config.AdminPasswords {
+		hdr := "Basic " + base64.StdEncoding.EncodeToString([]byte("admin:"+pw))
+		validAuthHeaders = append(validAuthHeaders, hdr)
+	}
+
+	// for paths that this middleware is applied to, enforce that the auth header must exist and match one of the known passwords
 	return func(c echo.Context) error {
 		hdr := c.Request().Header.Get("Authorization")
-		if hdr != headerVal {
+		if hdr == "" {
 			return echo.ErrForbidden
 		}
-		return next(c)
+		for _, val := range validAuthHeaders {
+			if subtle.ConstantTimeCompare([]byte(hdr), []byte(val)) == 1 {
+				return next(c)
+			}
+		}
+		return echo.ErrForbidden
```

Comment:
Bit of a nit, but since the user was not able to authenticate I'd call this a 401 Unauthorized.  As opposed to the user authenticated but does not have access (403 Forbidden).  Doing it by the book we'd also send a `WWW-Authenticate: Basic` back (technically a "must").

Reply:
(Another instance of this above when the authorization header is missing.)

Reply:
good catch!